File: 2024/19/solve.py
#!/usr/bin/env python3
import re
from functools import cache

line_iter = iter(open("input"))
patterns = next(line_iter).rstrip().split(", ")
next(line_iter)  # skip the blank line
lines = [line.rstrip() for line in line_iter]
count = sum(1 for line in lines if re.fullmatch(f'({"|".join(patterns)})+', line))


@cache
def n_ways(s: str) -> int:
    if not s:
        return 1
    return sum(
        n_ways(s[len(pattern) :]) for pattern in patterns if s.startswith(pattern)
    )


print(count)
total = sum(n_ways(s) for s in lines)
print(total)

File: 2024/19/solve.hs
#!/usr/bin/env runhaskell
{-# LANGUAGE BangPatterns #-}

import Control.DeepSeq (force)
import Data.Array (Array, array, (!))
import qualified Data.ByteString.Char8 as B
import Data.List (foldl')

type Pattern = B.ByteString

type Input = B.ByteString

-- Dynamic programming approach
countMatches :: [Pattern] -> Input -> Int
countMatches patterns input = dp ! n
  where
    n = B.length input
    dp = array (0, n) $ (0, 1) : [(i, ways i) | i <- [1 .. n]]

    ways :: Int -> Int
    ways i = sum [if matchAtPos p i then dp ! (i - B.length p) else 0 | p <- patterns]

    matchAtPos :: Pattern -> Int -> Bool
    matchAtPos p i
      | B.length p > i = False
      | otherwise = B.isPrefixOf p (B.drop (i - B.length p) input)

main :: IO ()
main = do
  input <- B.readFile "input"
  let (patternsLine : _ : cases) = B.lines input
      !patterns = force $ map B.strip $ B.split ',' patternsLine
      -- Pre-verify patterns are non-empty to avoid edge cases
      !validPatterns = filter (not . B.null) patterns
      !totalValidParses = foldl' (\acc s -> acc + countMatches validPatterns s) 0 cases
  print totalValidParses
